man/man1/rpslcheck.1
.\"// $Id$
.\"// Copyright (c) 2001,2002                        RIPE NCC
.\"//
.\"// All Rights Reserved
.\"//
.\"// Permission to use, copy, modify, and distribute this software and its
.\"// documentation for any purpose and without fee is hereby granted,
.\"// provided that the above copyright notice appear in all copies and that
.\"// both that copyright notice and this permission notice appear in
.\"// supporting documentation, and that the name of the author not be
.\"// used in advertising or publicity pertaining to distribution of the
.\"// software without specific, written prior permission.
.\"//
.\"// THE AUTHOR DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE, INCLUDING
.\"// ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS; IN NO EVENT SHALL
.\"// AUTHOR BE LIABLE FOR ANY SPECIAL, INDIRECT OR CONSEQUENTIAL DAMAGES OR ANY
.\"// DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN
.\"// AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\"// OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"//
.\"// 
.\"//  Copyright (c) 1994 by the University of Southern California
.\"//  and/or the International Business Machines Corporation.
.\"//  All rights reserved.
.\"//
.\"//  Permission to use, copy, modify, and distribute this software and
.\"//  its documentation in source and binary forms for lawful
.\"//  non-commercial purposes and without fee is hereby granted, provided
.\"//  that the above copyright notice appear in all copies and that both
.\"//  the copyright notice and this permission notice appear in supporting
.\"//  documentation, and that any documentation, advertising materials,
.\"//  and other materials related to such distribution and use acknowledge
.\"//  that the software was developed by the University of Southern
.\"//  California, Information Sciences Institute and/or the International
.\"//  Business Machines Corporation.  The name of the USC or IBM may not
.\"//  be used to endorse or promote products derived from this software
.\"//  without specific prior written permission.
.\"//
.\"//  NEITHER THE UNIVERSITY OF SOUTHERN CALIFORNIA NOR INTERNATIONAL
.\"//  BUSINESS MACHINES CORPORATION MAKES ANY REPRESENTATIONS ABOUT
.\"//  THE SUITABILITY OF THIS SOFTWARE FOR ANY PURPOSE.  THIS SOFTWARE IS
.\"//  PROVIDED "AS IS" AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES,
.\"//  INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
.\"//  MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE, TITLE, AND 
.\"//  NON-INFRINGEMENT.
.\"//
.\"//  IN NO EVENT SHALL USC, IBM, OR ANY OTHER CONTRIBUTOR BE LIABLE FOR ANY
.\"//  SPECIAL, INDIRECT OR CONSEQUENTIAL DAMAGES, WHETHER IN CONTRACT,
.\"//  TORT, OR OTHER FORM OF ACTION, ARISING OUT OF OR IN CONNECTION WITH,
.\"//  THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"//
.\"//  Questions concerning this software should be directed to 
.\"//  info-ra@isi.edu.
.\"//
.\"//  Author(s): Cengiz Alaettinoglu <cengiz@isi.edu>
.\"
.\"
.TH rpslcheck 1 local
.SH NAME
rpslcheck
.SH SYNOPSIS
.B rpslcheck
.RI [ flags ]\ [<as-no>]
.SH DESCRIPTION
.PP
rpslcheck syntax-checks the aut-num object for <as-no> registered 
in the Internet Routing Registry (IRR). If <as-no> is not specified,
object is expected as a standard input. 
Rpslcheck only checks the syntax of policy related attributes 
such as import, export and default.
If an aut-num passes rpslcheck's syntax checking,
it will not create cause errors to any of the tools in RAToolSet, 
such as RtConfig.
Future versions of this tool will generate better error messages and do
more than just syntax checking.
.SH IRRToolSet COMMON OPTIONS
.RS
.IP -help
Print summary of command-line options and abort.
.IP "\-T [whois_query | whois_response | input | all]"
Trace the argument. Useful for debugging.
.IP "\-D <integer>"
Print debugging messages of debug channel <integer> 
(if compiled with DEBUG defined).
.IP "\-version"
Print the version number and quit.
.IP "\-h <host-name>"
Establish an radbserver connection to host <host-name>.
The default is whois.radb.net
.IP "\-p <port-no>"
Establish an radbserver connection to port <port-no>.
The default is 43.
.IP "\-protocol <protocol>"
Use the <protocol> to connect to the IRR server. <protocol> can be irrd,
ripe, and bird. The default is irrd. The ripe option should be used with
the old ripe server.
.IP "\-s <source-list>"
Consider the sources specified in the comma separated <source-list>.
If an object is defined in multiple sources in <source-list>,
prcheck uses the definition first encountered in <source-list>
from left to right.
.IP \-rusage
Print resource usage upon completion.
.IP "\-ignore_errors"
Do not print error and warning messages due to communication to the
database server or parsing policy objects.
.IP "\-report_errors"
Print error and warning messages due to communication to the
database server or parsing policy objects.
.RE
.SH OTHER OPTIONS
.RS
.IP "\-as <as-no>"
Syntax check the aut-num object for <as-no>. 
If this option is not supplied, stdin is parsed.
.IP "\-echo"
Print the aut-num object before matching.
.RE
.SH ENVIRONMENT VARIABLES
.RS
.IP IRR_HOST
Specifies the irr host to connect.
.IP IRR_PORT
Specifies the irr port number to connect.
.IP IRR_SOURCES
Specifies the source list (comma separated) to consider.
.RE
.PP
Command line options take precedence over environment variables.
.SH EXAMPLES
.PP
.nf
	% ./rpslcheck 
	Congratulations no errors were found in parsing AS226	
	% rpslcheck AS3130
	Congratulations no errors were found in parsing AS3130
	% rpslcheck AS86
	Error: *it: AS701 192.41.177.6/32 192.41.177.249 (pref=MED) ANY
	Error:                            ^^^^^^^^^^^^^^
.fi
.PP
.SH ERROR AND WARNING MESSAGES
Error: no object for AS <as_no>
.RS
.PP
Aut-num object for <as_no> is not found in specified IRR for specified source. You could mistype the source, hostname or AS-number, or the object is truly non-existent.
.PP
.RE
For more error descriptions, please see
.B
Errors
manual page.
.SH AUTHORS
Cengiz Alaettinoglu <cengiz@isi.edu>

Katie Petrusha <katie@ripe.net>


NEWS
Release Notes for IRRToolSet
****************************

5.1.0 - 2016-04?-NN
------------------
* Starting with this release, IRRToolSet is going to follow a Semantic
  Versioning scheme: http://semver.org/
* Initial support for IOS XR config generation, use -config ciscoxr
* Rename @rtconfig set commands and add JunOS Support:
  cisco_prefix_acl_no to prefix_acl_no
  cisco_aspath_acl_no to aspath_acl_no
  cisco_pktfilter_acl_no to pktfilter_acl_no
  cisco_community_acl_no to cisco_community_acl_no
* Cisco IOS improvements:
  Move from peer-groups to peer templates
  Add support for community.delete()
  Reduced default starting id's for community lists, acl's etc from 100 to 50
  Added support for BGP communities starting with 0:
  Fix bug where both IPv4 & IPv6 prefixes would be added to the same prefix-list
* JunOS improvements:
  -junos_replace to output replace: configuration that can be loaded with "load replace"
  -junos_and_not_or to make route policies that AND community and prefix-lists to match
   the rtconfig Cisco behaviour.
  Increment JunOS policy-statement id's for IPv4/v6 to avoid duplicate id's
  Add support for IPv6 prefix-length-range to bring it to parity with IPv4
  Fixed minor configuration output ordering
  Fix problems with 4-byte AS's

5.0.1
-----
* don't barf on pingable: and ping-hdl: attributes (RFC 5943)
* don't barf on inet6num: assignment-size (RIPE policy proposal 2010-06)
* RFC 5952 requires that ipv6 addresses be presented in lower case.
* ipv6 addresses are now printed in optimised format.
* added geoloc: and language: fields to both inetnum: and inet6num: objects.
* compilation fixes for g++ 4.7 / debian wheeze
* stop talking asdot to IRRDBs.  none of the IRRs supports asdot since ~2010
* silence a bunch of warnings for clang 3.0
* fix AS output for rpslcheck

5.0.0
-----
Substantial reorganisation of package.  Only rtconfig, peval and rpslcheck survived. 
For further information on changes for 5.0.0, please consult the irrtoolset
wiki on http://irrtoolset.isc.org/

Moreover, IRRToolSet was switched to a new MIT style license, from a
non-commercial license previously. Thanks to past author Cengiz Alaettinoglu
and partial copyright holder, University of Southern California, for graciously
arranging this.

Miscellanous fixes:
* fix f_junos.cc to make export and import token names consistent. Thanks to
  Stan Barber.
* Support for 4 byte ASNs. Many thanks to S.P.Zeidler for this.
* Makefile now honors DESTDIR. Thanks to Kaj Niemi for the patch.
* @RtConfig fix for IPv6. Thanks to Nick Hilliard for the report and
  to Timo Koehler for the patch.
* Fix when an IOS as-path list from an RPSL as-set contains more
  than one AS number. Thanks to Hagen Boehm for the patch.
* Compile fix for Solaris 9 and other systems that do not have
  stdint.h defined. Thanks to Hagen Boehm for reporting the problem.
* Fix for core dump on Solaris. Thanks to Matthew Moyle-Croft for the
  report and to S.P.Zeidler for the fix.
* Fix for invalid RtConfig output on Cisco when specifying family.
  Thanks to Daryl Collins for the report and to S.P.Zeidler for the
  fix.
* Cache file support when "irrd" protocol is used. Thanks to Hagen
  Boehm for the implementation.
* Build fixes for OS X.
* Build fix for gcc 4.2 (default on Debian and Ubuntu these days).
* Explicit check for flex and bison in configure.

4.8.5
-----
Bug fix for RPSLng queries on IRRd servers.

Better support for 64-bit compilation.

Integrated a number of fixes from the NetBSD port as proposed by 
Havard Eidnes.

* Numerous compiler fixes:
  o src/config.h.in is amended to conditionally define __P(x).  Use
    __P() instead of _G_ARGS() in various places.
  o Many, many files are patched to replace occurrances of _G_uint32_t
    with uint32_t. The former is in the host compiler implementation's
    name space, and as far as I know, a library writer has no business
    assuming anything about the types defined in that name space
    (names starting with a single underscore character).
  o Instead of including <_G_config.h>, use <config.h>, and add
    includes of e.g. <sys/types.h> where needed.
  o A few troublesome casts removed which newer C++ compilers balk at.

* JunOS support enabled & fixed

* Portability to Dragonfly BSD (treat as FreeBSD)

* Bugfix for Cisco export

4.8.4
-----
* A number of runtime problems which caused crashes were reported, and
  these have been fixed. Thanks to Hagen Boehm for discovering them.

4.8.3
-----
A set of patches collected by Hagen Boehm was applied:

* whois.apnic.net server problem
  APNIC's whois server (RIPE NCC software with RPSS functions
  disabled) has changed syntax of some embedded comments in a way that
  violates the protocol assumptions in IRRToolSet 4.8.2
  (required/allowed "empty lines" ~ "\n\n" in the context of "%"
  comment lines. This problem seems to be around since APNIC switched
  to a database version supporting IPv6/RPSLng).
  Thanks to Hagen Boehm.

* ASnnn embedded in AS-Set AS-xxxx doesn't resolve properly (prefix
  loss)
  We are seeing cases where resolving to route-sets (a) an ASnnn and
  (b) an AS-set AS-xxxx which includes ASnnn as a member lists some
  prefixes as members of ASnnn but NOT as members of AS-xxxx (Which
  turns something like "125.16.0.0/14^14-24" into some thousand
  separately listed 125.16.x.y/n^n-n or 125.16.x.y/n^m-m that don't
  combine into any 125.16.x.y/n^m-l unless m=l).
  Thanks to Hagen Boehm.

* multi protocol filter problem 
  prefix filter consisting of ipv4 and ipv6 prefixes get not resolved
  properly.  Either only prefixes of a single protocol get resolved
  (e.g. only ipv4) or the route-set contains ANY. The first case was
  observed when resolving an "ASxxxx" filter, the latter case when
  resolving an "ASxxxx ASyyyy" filter, where both ASxxxx and ASyyyy
  contained ipv4 and ipv6 prefixes.
  Thanks to Hagen Boehm.

* due to an uninitialized integer the irrd client showed erroneous

trunk/src/rpslcheck/rpslcheck.cc
//  $Id$
// Copyright (c) 2001,2002                        RIPE NCC
//
// All Rights Reserved
//
// Permission to use, copy, modify, and distribute this software and its
// documentation for any purpose and without fee is hereby granted,
// provided that the above copyright notice appear in all copies and that
// both that copyright notice and this permission notice appear in
// supporting documentation, and that the name of the author not be
// used in advertising or publicity pertaining to distribution of the
// software without specific, written prior permission.
//
// THE AUTHOR DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE, INCLUDING
// ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS; IN NO EVENT SHALL
// AUTHOR BE LIABLE FOR ANY SPECIAL, INDIRECT OR CONSEQUENTIAL DAMAGES OR ANY
// DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN
// AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
// OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
//
//
//  Copyright (c) 1994 by the University of Southern California
//  All rights reserved.
//
//  Permission to use, copy, modify, and distribute this software and its
//  documentation in source and binary forms for lawful non-commercial
//  purposes and without fee is hereby granted, provided that the above
//  copyright notice appear in all copies and that both the copyright
//  notice and this permission notice appear in supporting documentation,
//  and that any documentation, advertising materials, and other materials
//  related to such distribution and use acknowledge that the software was
//  developed by the University of Southern California, Information
//  Sciences Institute. The name of the USC may not be used to endorse or
//  promote products derived from this software without specific prior
//  written permission.
//
//  THE UNIVERSITY OF SOUTHERN CALIFORNIA DOES NOT MAKE ANY
//  REPRESENTATIONS ABOUT THE SUITABILITY OF THIS SOFTWARE FOR ANY
//  PURPOSE.  THIS SOFTWARE IS PROVIDED "AS IS" AND WITHOUT ANY EXPRESS OR
//  IMPLIED WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED
//  WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE,
//  TITLE, AND NON-INFRINGEMENT.
//
//  IN NO EVENT SHALL USC, OR ANY OTHER CONTRIBUTOR BE LIABLE FOR ANY
//  SPECIAL, INDIRECT OR CONSEQUENTIAL DAMAGES, WHETHER IN CONTRACT, TORT,
//  OR OTHER FORM OF ACTION, ARISING OUT OF OR IN CONNECTION WITH, THE USE
//  OR PERFORMANCE OF THIS SOFTWARE.
//
//  Questions concerning this software should be directed to 
//  ratoolset@isi.edu.
//
//  Author(s): Cengiz Alaettinoglu <cengiz@ISI.EDU>
//	       Katie Petrusha <katie@ripe.net>

#include <config.h>
#include <iostream>
#include "rpsl/object.hh"
#include "irrutil/rusage.hh"
#include "irrutil/debug.hh"
#include "irrutil/trace.hh"
#include "irrutil/Argv.hh"
#include "irrutil/version.hh"
#include "irr/irr.hh"
#include "irr/rawhoisc.hh"
#include "rpsl/schema.hh"

using namespace std;

Rusage ru;
bool opt_stats                   = false;
bool opt_rusage                  = false;
char *opt_prompt                 = (char *)"rpslcheck> ";
bool opt_echo                    = false;
bool opt_asdot                   = false;
char *opt_my_as			 = NULL;
#ifdef ENABLE_DEBUG
bool opt_debug_rpsl              = false;
#endif // ENABLE_DEBUG

int start_tracing(char *dst, char *key, char *nextArg) {
   if (nextArg) {
      trace.enable(nextArg);
      return 1; // return 1 to signify nextArg is used by us
   }
   return 0; 
}

int start_debugging(char *dst, char *key, char *nextArg) {
   if (nextArg) {
      Debug(dbg.enable(atoi(nextArg)));
      return 1; // return 1 to signify nextArg is used by us
   }
   return 0;
}

void init_and_set_options (int argc, char **argv, char **envp) {
   ArgvInfo argTable[] = {
     // RAToolSet common arguments
     // key, type, src, dst, help
     {"-T", ARGV_FUNC, (char *) &start_tracing,      (char *) NULL, 
      "Start tracing the next argument"},
     {"-D", ARGV_FUNC, (char *) &start_debugging,    (char *) NULL, 
      "Start debugging the next argument"},
     {"-version", ARGV_FUNC, (char *) &version,      (char *) NULL,
      "Show version"},
     
     IRR_COMMAND_LINE_OPTIONS,

     {"-asdot", ARGV_BOOL, (char *) NULL, (char *) &opt_asdot,
      "print AS numbers in asdot format."},

     {"-rusage", ARGV_BOOL, (char *) NULL,           (char *) &opt_rusage,
      "On termination print resource usage"},
     {"-stats", ARGV_BOOL, (char *) NULL,            (char *) &opt_stats,
      "On termination print class statistics"},
     {"-prompt", ARGV_STRING,  (char *) NULL,        (char *) &opt_prompt,
      "Prompt"},
     {"-as",  ARGV_STRING,     (char *) NULL, (char *) &opt_my_as, 
      "AS number of the aut-num object to parse"},
     
     {"-echo", ARGV_BOOL, (char *) NULL,           (char *) &opt_echo,
      "Echo each object parsed"},
#ifdef ENABLE_DEBUG
     {"-debug_rpsl", ARGV_BOOL, (char *) NULL,     (char *) &opt_debug_rpsl,
      "Turn on bison debugging. Intended for developers."},
#endif // ENABLE_DEBUG

     {(char *) NULL, ARGV_END, (char *) NULL, (char *) NULL,
      (char *) NULL}
   };
  
   IRR::handleEnvironmentVariables(envp);

   if (ParseArgv(&argc, argv, argTable, ARGV_NO_LEFTOVERS) != ARGV_OK) {
      std::cerr << std::endl;
      exit(1);
   }

   if (! (irr = IRR::newClient())) {
      std::cerr << "Unknown protocol!" << std::endl;
      exit(1);
   }

   // have a prompt only if the input is coming from a tty
   if (!isatty(fileno(stdin)) || !isatty(fileno(stdout)))
      opt_prompt = NULL;
}


int main(int argc, char **argv, char **envp) {
   schema.initialize();
   init_and_set_options(argc, argv, envp);
   schema.beHarsh();

   // opt_echo = 1;

#ifdef ENABLE_DEBUG
   extern int rpsldebug;
   if (opt_debug_rpsl)
      rpsldebug = 1;
#endif // ENABLE_DEBUG

   Object *o;
   bool code = true;
   int myAS;
   
   while (opt_my_as || cin ) {
       if (opt_my_as) {
          // if the first two characters of the ASN are "as", then ignore them
          if (strcasestr(opt_my_as, "as") == opt_my_as)
            opt_my_as += 2;

          const char *dot = strchr(opt_my_as,'.');
          if (dot)
             myAS = atoi(opt_my_as)<<16 | atoi(dot+1);
          else
	     myAS = atoi(opt_my_as);
	  const AutNum *autnum = irr->getAutNum(myAS);
          if (!autnum)	{
          	std::cerr << "Error: no object for AS " << myAS << std::endl;
		exit(1);
	  }
	  o = new Object((Object &) *autnum);
          o->scan();
       }
       else {
	  o = new Object;
          code = o->read();
          if (code)
             code = o->scan();
       }

          if (opt_echo && code)
             std::cout << *o;

          if (!o->has_error && code)
          std::cout << "Congratulations no errors were found" << std::endl;

	  delete o;

	  if (opt_my_as)	{
		break;
	  }

   }

   if (opt_stats)
      schema.printClassStats();

   if (opt_prompt)
      std::cout << std::endl;

   if (opt_rusage)
      clog << ru;
}
